### ISSUE TITLE
Remove Python 2.7

### ISSUE THREAD

**Issue by bedroge** [EESSI/compatibility-layer#36]
We should try to remove Python 2 from our compatibility layer.

**Comment by peterstol** [comment]
The default is python3. There is stil quite a list with dependencies. Better to wait a bit.
```
 equery d dev-lang/python-2.7.18-r1:2.7
 * These packages depend on dev-lang/python-2.7.18-r1:2.7:
app-crypt/gpgme-1.14.0 (python_targets_python3_6 ? dev-lang/python:3.6)
                       (python_targets_python3_7 ? dev-lang/python:3.7)
                       (python_targets_python3_8 ? dev-lang/python:3.8)
app-eselect/eselect-repository-7 (python_single_target_python3_6 ? dev-lang/python:3.6)
                                 (python_single_target_python3_7 ? dev-lang/python:3.7)
                                 (python_single_target_python3_8 ? dev-lang/python:3.8)
app-misc/ca-certificates-20200601.3.53 (dev-lang/python:3.8)
                                       (dev-lang/python:3.7)
                                       (dev-lang/python:3.6)
app-misc/pax-utils-1.2.6 (python_single_target_python3_6 ? dev-lang/python:3.6)
                         (python_single_target_python3_7 ? dev-lang/python:3.7)
                         (python_single_target_python3_8 ? dev-lang/python:3.8)
app-portage/gentoolkit-0.5.0 (python_targets_python3_6 ? dev-lang/python:3.6[xml(+),threads(+)])
                             (python_targets_python3_7 ? dev-lang/python:3.7[xml(+),threads(+)])
                             (python_targets_python3_8 ? dev-lang/python:3.8[xml(+),threads(+)])
dev-libs/libxml2-2.9.10-r1 (python_targets_python2_7 ? >=dev-lang/python-2.7.5-r2:2.7[xml])
                           (python_targets_python3_6 ? dev-lang/python:3.6[xml])
                           (python_targets_python3_7 ? dev-lang/python:3.7[xml])
                           (python_targets_python3_8 ? dev-lang/python:3.8[xml])
                           (python_targets_python3_9 ? dev-lang/python:3.9[xml])
dev-python/certifi-10001 (python_targets_python2_7 ? >=dev-lang/python-2.7.5-r2:2.7)
                         (python_targets_python3_6 ? dev-lang/python:3.6)
                         (python_targets_python3_7 ? dev-lang/python:3.7)
                         (python_targets_python3_8 ? dev-lang/python:3.8)
                         (python_targets_python3_9 ? dev-lang/python:3.9)
dev-python/click-7.1.2 (>=dev-lang/python-2.7.5-r2:2.7)
                       (dev-lang/python:3.7)
dev-python/cython-0.29.21 (python_targets_python2_7 ? >=dev-lang/python-2.7.5-r2:2.7[threads(+)])
                          (python_targets_python3_6 ? dev-lang/python:3.6[threads(+)])
                          (python_targets_python3_7 ? dev-lang/python:3.7[threads(+)])
                          (python_targets_python3_8 ? dev-lang/python:3.8[threads(+)])
                          (python_targets_python3_9 ? dev-lang/python:3.9[threads(+)])
                          (doc ? dev-lang/python:3.9[threads(+)])
                          (dev-lang/python:3.8[threads(+)])
                          (dev-lang/python:3.7[threads(+)])
                          (dev-lang/python:3.6[threads(+)])
                          (>=dev-lang/python-2.7.5-r2:2.7[threads(+)])
dev-python/lxml-4.5.2 (python_targets_python2_7 ? >=dev-lang/python-2.7.5-r2:2.7)
                      (python_targets_python3_6 ? dev-lang/python:3.6)
                      (python_targets_python3_7 ? dev-lang/python:3.7)
                      (python_targets_python3_8 ? dev-lang/python:3.8)
                      (python_targets_python3_9 ? dev-lang/python:3.9)
dev-python/setuptools-46.4.0-r2 (>=dev-lang/python-2.7.5-r2:2.7[xml(+)])
                                (dev-lang/python:3.7[xml(+)])
dev-python/setuptools_scm-4.1.2 (>=dev-lang/python-2.7.5-r2:2.7)
                                (dev-lang/python:3.7)
dev-python/six-1.15.0 (python_targets_python2_7 ? >=dev-lang/python-2.7.5-r2:2.7)
                      (python_targets_python3_6 ? dev-lang/python:3.6)
                      (python_targets_python3_7 ? dev-lang/python:3.7)
                      (python_targets_python3_8 ? dev-lang/python:3.8)
                      (python_targets_python3_9 ? dev-lang/python:3.9)
dev-vcs/git-2.28.0 (python_single_target_python3_6 ? dev-lang/python:3.6)
                   (python_single_target_python3_7 ? dev-lang/python:3.7)
                   (python_single_target_python3_8 ? dev-lang/python:3.8)
net-misc/wget-1.20.3-r3 (test ? dev-lang/python:3.8)
                        (test ? dev-lang/python:3.7)
                        (test ? dev-lang/python:3.6)
sys-apps/archspec-0.1.1 (python_targets_python3_6 ? dev-lang/python:3.6)
                        (python_targets_python3_7 ? dev-lang/python:3.7)
                        (python_targets_python3_8 ? dev-lang/python:3.8)
                        (python_targets_python3_9 ? dev-lang/python:3.9)
sys-apps/coreutils-8.32-r1 (test ? dev-lang/python:3.8)
                           (test ? dev-lang/python:3.7)
                           (test ? dev-lang/python:3.6)
sys-apps/findutils-4.7.0 (test ? dev-lang/python:3.9)
                         (test ? dev-lang/python:3.8)
                         (test ? dev-lang/python:3.7)
                         (test ? dev-lang/python:3.6)
sys-apps/portage-3.0.4-r1 (python_targets_python3_6 ? dev-lang/python:3.6[ssl(+)])
                          (python_targets_python3_7 ? dev-lang/python:3.7[ssl(+)])
                          (python_targets_python3_8 ? dev-lang/python:3.8[ssl(+)])
                          (python_targets_python3_9 ? dev-lang/python:3.9[ssl(+)])
                          (python_targets_python3_6 ? dev-lang/python:3.6[bzip2(+),threads(+)])
                          (python_targets_python3_7 ? dev-lang/python:3.7[bzip2(+),threads(+)])
                          (python_targets_python3_8 ? dev-lang/python:3.8[bzip2(+),threads(+)])
                          (python_targets_python3_9 ? dev-lang/python:3.9[bzip2(+),threads(+)])
sys-apps/util-linux-2.36 (python_targets_python3_6 ? dev-lang/python:3.6)
                         (python_targets_python3_7 ? dev-lang/python:3.7)
                         (python_targets_python3_8 ? dev-lang/python:3.8)
sys-devel/automake-1.16.2 (test ? dev-lang/python:3.7)
                          (test ? dev-lang/python:3.6)
sys-libs/glibc-2.32 (dev-lang/python:3.8)
                    (dev-lang/python:3.7)
                    (dev-lang/python:3.6)
sys-libs/libseccomp-2.5.0 (python_targets_python3_6 ? dev-lang/python:3.6)
                          (python_targets_python3_7 ? dev-lang/python:3.7)
                          (python_targets_python3_8 ? dev-lang/python:3.8)
                          (python_targets_python3_9 ? dev-lang/python:3.9)
```

**Comment by bedroge** [comment]
I wonder if they all really depend on Python 2, though, since they all have a Python 3 target as well. I started a Prefix installation this afternoon with `PYTHON_TARGETS="python3_7"`, which just completed, and that doesn't have Python 2 anymore (only a symlink which doesn't really do anything). The dependency for, for instance, `eselect-repository` is probably just there because it can be built with Python 2 (and Python 2 is still available in the repo), but doesn't actually need to be built with it (as long as you have another compatible Python version).

The default behavior is to install Python 2:
```
$ emerge --ask eselect-repository
These are the packages that would be merged, in order:

Calculating dependencies... done!
[ebuild  N     ] app-arch/unzip-6.0_p25-r1  USE="bzip2 unicode -natspec" 
[ebuild  N     ] dev-libs/libxslt-1.1.34-r1  USE="crypt -debug -examples -static-libs" 
[ebuild  NS    ] dev-lang/python-2.7.18-r3 [3.7.9] USE="(threads%*) (wide-unicode%*) (-berkdb)" 
[ebuild  N     ] dev-python/certifi-10001  USE="-test" PYTHON_TARGETS="python2_7 python3_7 -pypy3 -python3_6 -python3_8 -python3_9" 
[ebuild  N     ] dev-python/setuptools-46.4.0-r3  USE="-test" PYTHON_TARGETS="python2_7 python3_7 -pypy3 -python3_6 -python3_8 -python3_9" 
[ebuild  N     ] dev-python/setuptools_scm-4.1.2-r1  USE="-test" PYTHON_TARGETS="python3_7 -pypy3 -python3_6 -python3_8 -python3_9" 
[ebuild  N     ] dev-python/cython-0.29.21  USE="-doc -emacs -test" PYTHON_TARGETS="python2_7 python3_7 -pypy3 -python3_6 -python3_8 -python3_9" 
[ebuild  N     ] dev-python/lxml-4.5.2  USE="threads -doc -examples -test" PYTHON_TARGETS="python2_7 python3_7 -pypy3 -python3_6 -python3_8 -python3_9" 
[ebuild  N     ] app-eselect/eselect-repository-8  PYTHON_SINGLE_TARGET="python3_7 -python3_6 -python3_8" 

```

But:
```
$ env PYTHON_TARGETS="python3_7" emerge --ask eselect-repository

These are the packages that would be merged, in order:

Calculating dependencies... done!
[ebuild  N     ] dev-python/certifi-10001  USE="-test" PYTHON_TARGETS="python3_7 -pypy3 -python2_7 -python3_6 -python3_8 -python3_9" 
[ebuild  N     ] app-arch/unzip-6.0_p25-r1  USE="bzip2 unicode -natspec" 
[ebuild  N     ] dev-libs/libxslt-1.1.34-r1  USE="crypt -debug -examples -static-libs" 
[ebuild  N     ] dev-python/setuptools-46.4.0-r3  USE="-test" PYTHON_TARGETS="python3_7 -pypy3 -python2_7 -python3_6 -python3_8 -python3_9" 
[ebuild  N     ] dev-python/setuptools_scm-4.1.2-r1  USE="-test" PYTHON_TARGETS="python3_7 -pypy3 -python3_6 -python3_8 -python3_9" 
[ebuild  N     ] dev-python/cython-0.29.21  USE="-doc -emacs -test" PYTHON_TARGETS="python3_7 -pypy3 -python2_7 -python3_6 -python3_8 -python3_9" 
[ebuild  N     ] dev-python/lxml-4.5.2  USE="threads -doc -examples -test" PYTHON_TARGETS="python3_7 -pypy3 -python2_7 -python3_6 -python3_8 -python3_9" 
[ebuild  N     ] app-eselect/eselect-repository-8  PYTHON_SINGLE_TARGET="python3_7 -python3_6 -python3_8" 
```

So, I think that most things (maybe even everything) should work if we put that variable in `/etc/portage/make.conf`.

**Comment by bedroge** [comment]
This page also has some useful information:
https://wiki.gentoo.org/wiki/Project:Python/PYTHON_TARGETS

**Comment by bedroge** [comment]
Tested this a bit more by installing our overlay and packages, and that worked fine.

```
$ ls -ld usr/lib/python*
drwxr-xr-x  3 bob bob 4096 Sep 23 15:18 usr/lib/python-exec
drwxr-xr-x 32 bob bob 4096 Sep 23 16:11 usr/lib/python3.7

$ ls -l usr/bin/python*
lrwxrwxrwx 1 bob bob    13 Sep 23 15:18 usr/bin/python -> python-exec2c
lrwxrwxrwx 1 bob bob    31 Sep 23 15:18 usr/bin/python-config -> ../lib/python-exec/python-exec2
-rwxr-xr-x 1 bob bob 18672 Sep 23 15:18 usr/bin/python-exec2c
lrwxrwxrwx 1 bob bob    13 Sep 23 15:18 usr/bin/python2 -> python-exec2c
lrwxrwxrwx 1 bob bob    31 Sep 23 15:18 usr/bin/python2-config -> ../lib/python-exec/python-exec2
lrwxrwxrwx 1 bob bob    13 Sep 23 15:18 usr/bin/python3 -> python-exec2c
lrwxrwxrwx 1 bob bob    31 Sep 23 15:18 usr/bin/python3-config -> ../lib/python-exec/python-exec2
lrwxrwxrwx 1 bob bob    10 Sep 23 16:11 usr/bin/python3.7 -> python3.7m
lrwxrwxrwx 1 bob bob    17 Sep 23 16:11 usr/bin/python3.7-config -> python3.7m-config
-rwxr-xr-x 1 bob bob 14224 Sep 23 16:11 usr/bin/python3.7m
-rwxr-xr-x 1 bob bob  2785 Sep 23 16:11 usr/bin/python3.7m-config

$ python2
python2: no python-exec wrapped executable found in /local/tmp/gentoo2/usr/lib/python-exec.

$ python3
Python 3.7.9 (default, Sep 23 2020, 16:10:25) 
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
```

**Comment by boegel** [comment]
I propose we try this for the `2020.09` version of the pilot repo, before we start building stuff in the software layer?

**Comment by bedroge** [comment]
I just published a new installation without Python2 in `/cvmfs/pilot.eessi-hpc.org/2020.09/compat/x86_64/`. The overlay/packages/fixes have not been installed yet. In order to do this, we first have to decide how/where we are going to set the flags for not building for Python2:
```
*/* PYTHON_TARGETS: -python2_7
*/* PYTHON_SINGLE_TARGET: -* python3_7
```

This can be done in a file `etc/portage/package.use`, or in a file (e.g. `disable_python2`) in a `etc/portage/package.use` directory,  or by adding something to the file `etc/portage/make.conf`, and probably in more ways... ;-)
I would prefer the second option, what do you think?

One other thing related to this: our fork of CC's gentoo-overlay repo now contains their configuration files in https://github.com/EESSI/gentoo-overlay/tree/master/etc/portage, except for the sets dir that we added. Maybe it makes sense to store our configuration files there (e.g. the `package.use` file/directory), so what do we do with the current ones? Wipe them all?

**Comment by boegel** [comment]
I would go with a file named `disable_python2` in `etc/portage/package.use` too, that seems most transparent to me.

W.r.t. the CC configuration files: are those picked up automatically, or do we need to opt-in using them somehow?

I'm in favor of removing stuff we're not sure about if/why we need it, we can add stuff back later gradually as the need arises.

**Comment by bedroge** [comment]
They're not picked up automatically: just like with the `sets` dir, you have to copy them to `$EPREFIX/etc/portage` first.

**Comment by boegel** [comment]
Then it doesn't hurt to leave them in the repo?

**Comment by bedroge** [comment]
Then it might become a bit unclear what is what? If we want to go for a `package.use` dir, we already have to remove their `package.use` file.

We could even just copy that entire directory from our repo to `etc/portage` in our installation script/playbook, but then we definitely need to wipe it first.

**Comment by boegel** [comment]
OK, let's wipe it then. It's not like we're permanently losing something, it's all in git history anyway...

**Comment by peterstol** [comment]
Ok, then we agree on unmasking python3 and can close this issue.

**Comment by bedroge** [comment]
Just realized that there is one downside, see https://github.com/EESSI/gentoo-overlay/pull/17:

One thing that might be slightly annoying about having these config files in our overlay: in order to install our overlay, we need to emerge some packages, e.g. eselect-repository. And these might install Python2 if we are not careful. So we will have to set some variables while installing those packages.

**Comment by bedroge** [comment]
Implemented in #40.
